### PR TITLE
DEBUG: Allow debug levels to be specified without needing to recompile

### PR DESCRIPTION
**Pull request by zathras-crypto**
This small PR allows debug levels to be set as startup parameters.

### CODE REVIEW

**Comment by dexX7** [comment]
This is nice!

Maybe we can move that block into `DebugLogInit()` in `src/omnicore/log.cpp` below `mutexDebugLog = new boost::mutex();` to have everything together?

```diff
diff --git a/src/omnicore/log.cpp b/src/omnicore/log.cpp
index d4882a5..9d3ccdc 100644
--- a/src/omnicore/log.cpp
+++ b/src/omnicore/log.cpp
@@ -12,6 +12,7 @@
 #include <assert.h>
 #include <stdio.h>
 #include <string>
+#include <vector>
 
 // Log files
 const std::string LOG_FILENAME    = "mastercore.log";
@@ -81,6 +82,38 @@ static void DebugLogInit()
     if (fileout) setbuf(fileout, NULL); // Unbuffered
 
     mutexDebugLog = new boost::mutex();
+
+    // when providing support recompiling to enable debug levels is not always feasible
+    // determine whether to override debug levels via enumerating startup option --omnidebug
+    // example usage: --omnidebug=parser --omnidebug=metadex1 --omnidebug=ui
+    if (!mapMultiArgs["-omnidebug"].empty()) {
+        std::vector<std::string>& debugLevels = mapMultiArgs["-omnidebug"];
+        for (std::vector<std::string>::const_iterator it = debugLevels.begin(); it != debugLevels.end(); ++it) {
+            if (*it == "parser_data") msc_debug_parser_data = true;
+            if (*it == "parser") msc_debug_parser = true;
+            if (*it == "verbose") msc_debug_verbose = true;
+            if (*it == "verbose2") msc_debug_verbose2 = true;
+            if (*it == "verbose3") msc_debug_verbose3 = true;
+            if (*it == "vin") msc_debug_vin = true;
+            if (*it == "script") msc_debug_script = true;
+            if (*it == "dex") msc_debug_dex = true;
+            if (*it == "send") msc_debug_send = true;
+            if (*it == "tokens") msc_debug_tokens = true;
+            if (*it == "spec") msc_debug_spec = true;
+            if (*it == "exo") msc_debug_exo = true;
+            if (*it == "tally") msc_debug_tally = true;
+            if (*it == "sp") msc_debug_sp = true;
+            if (*it == "sto") msc_debug_sto = true;
+            if (*it == "txdb") msc_debug_txdb = true;
+            if (*it == "tradedb") msc_debug_tradedb = true;
+            if (*it == "persistence") msc_debug_persistence = true;
+            if (*it == "ui") msc_debug_ui = true;
+            if (*it == "pending") msc_debug_pending = true;
+            if (*it == "metadex1") msc_debug_metadex1 = true;
+            if (*it == "metadex2") msc_debug_metadex2 = true;
+            if (*it == "metadex3") msc_debug_metadex3 = true;
+        }
+    }
 }
 
 /**
```

**Comment by zathras-crypto** [comment]
> Maybe we can move that block

Yeah I didn't think of that - I'm so used to sticking startup params in ```mastercore_init()``` - I had a couple of updates I wanted to make so I'll move that now in with those changes...

Also, perhaps time to change mastercore.log to omnicore.log?

**Comment by dexX7** [comment]
> Also, perhaps time to change mastercore.log to omnicore.log?

Yep, seems very reasonable. Let me check out, if we can make it configurable, too. I'll push an update to `0.0.10-Z-Debug` later.

Regarding the categories: at some point we may also re-evaluate using the logging style Bitcoin Core uses, with string categories:

```c++
if (msc_debug_metadex1) { file_log("log message\n"; }
```
vs.
```c++
PrintToLog("metadex1", "log message\n");
```

I'll look into it, if you think this is something we should do rather sooner than later.

**Comment by dexX7** [comment]
Sorry for the back and forth, I noticed setting the log levels via `DebugLogInit()` doesn't cover the case where no log file is used, for example, when using `--printtoconsole`.

I moved the explicit setup back into `mastercore_init()`, but wrapped it into `InitDebugLogLevels()` in `log.cpp`.

By the way, do we still need `mastercore_crowdsales.log` or `mastercore_owners.log`? To my knowledge this was only used for debug purposes during the initial implementation, and I'd prefer to get rid of them - unless they have some other purpose of course.

**Comment by zathras-crypto** [comment]
> By the way, do we still need mastercore_crowdsales.log or mastercore_owners.log?

I don't believe so - this was down to those initial debates in the early days (I was pushing for exposing the data via RPC but there was some resistance due to workload and timeframes which led to this kind of flat file that could be scraped via scripting) - I really didn't like that so I was a bit of a sh*t and just went ahead and coded up the needed RPC calls 'on the side' so I'm pretty sure those files are redundant now.

TL:DR; I believe the data in those files can be obtained over RPC now but we should probably just double check (it was a while ago!) just to be sure.

**Comment by dexX7** [comment]
> I don't believe so

Alright, then let's keep it in mind, and tackle it at some point later.
